src/main/java/org/opensearch/knn/index/query/iterators/GroupedNestedDocIdSetIterator.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.index.query.iterators;
+
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.util.BitSet;
+
+import java.io.IOException;
+import java.util.ArrayList;
+
+/**
+ * A `DocIdSetIterator` that iterates over all nested document IDs belongs to the same parent document for a given
+ * set of nested document IDs.
+ *
+ * The {@link #docIds} should include only a single nested document ID per parent document. Otherwise, the nested documents
+ * of that parent document will be iterated multiple times.
+ *
+ * The order of document IDs in {@link #docIds} does not need to be sorted.
+
+ */
+public class GroupedNestedDocIdSetIterator extends DocIdSetIterator {
+    private final BitSet parentBitSet;
+    private final ArrayList<Integer> docIds;
+    private Long cost;
+    private int currentIndex;
+    private int currentDocId;
+    private int currentParentId;
+
+    public GroupedNestedDocIdSetIterator(final BitSet parentBitSet, final ArrayList<Integer> docIds) {
+        this.parentBitSet = parentBitSet;
+        this.docIds = docIds;
+        currentIndex = -1;
+        currentDocId = -1;
+    }
+
+    @Override
+    public int docID() {
+        return currentDocId;
+    }
+
+    @Override
+    public int nextDoc() throws IOException {
+        if (currentDocId == NO_MORE_DOCS) {
+            return currentDocId;
+        }
+
+        if (currentDocId == -1) {
+            moveToNextIndex();
+            return currentDocId;
+        }
+
+        currentDocId++;
+        assert currentDocId <= currentParentId;
+        if (currentDocId == currentParentId) {
+            moveToNextIndex();
+        }
+        return currentDocId;
+    }
+
+    @Override
+    public int advance(final int i) throws IOException {
+        if (currentDocId == NO_MORE_DOCS) {
+            return currentDocId;
+        }
+
+        return slowAdvance(i);
+    }
+
+    @Override
+    public long cost() {
+        if (cost == null) {
```

Comment:
do you think it will be better to go with `cost == -1` and use primitives rather than `cost == null`